lib_neutral_prompt/cfg_denoiser_hijack.py

```diff
@@ -212,27 +212,110 @@ def filter_abs_top_k(vector: torch.Tensor, k_ratio: float) -> torch.Tensor:
+
+        model = self.inner_model.inner_model.forge_objects.unet.model
+        control = self.inner_model.inner_model.forge_objects.unet.controlnet_linked_list
+        extra_concat_condition = self.inner_model.inner_model.forge_objects.unet.extra_concat_condition
+        x = denoiser_params.x
+        timestep = denoiser_params.sigma
+        model_options = self.inner_model.inner_model.forge_objects.unet.model_options
+        seed = self.p.seeds[0]
+
+        uncond = forge_sampler.cond_from_a1111_to_patched_ldm(denoiser_params.text_uncond)
+        conds = forge_sampler.cond_from_a1111_to_patched_ldm_weighted(denoiser_params.text_cond, cond_composition)
+        conds = conds + uncond
+
+        denoised = []
+
+        for current_cond in conds:
+            cond = [current_cond]
+            cond[0]['strength'] = 1.0
+
+            if extra_concat_condition is not None:
+                image_cond_in = extra_concat_condition
+            else:
+                image_cond_in = denoiser_params.image_cond
+
+            if isinstance(image_cond_in, torch.Tensor):
+                if image_cond_in.shape[0] == x.shape[0] \
+                        and image_cond_in.shape[2] == x.shape[2] \
+                        and image_cond_in.shape[3] == x.shape[3]:
+                    cond[0]['model_conds']['c_concat'] = CONDRegular(image_cond_in)
+
+            if control is not None:
+                cond[0]['control'] = control
+
+            for modifier in model_options.get('conditioning_modifiers', []):
+                model, x, timestep, _, cond, cond_scale, model_options, seed = modifier(model, x, timestep, None, cond, cond_scale, model_options, seed)
+
+            model_options["disable_cfg1_optimization"] = True
+
+            result = sampling_function(model, x, timestep, None, cond, 1.0, model_options, seed)
+            denoised.append(result)
 
-    sampler.model_wrap_cfg.combine_denoised = functools.partial(
-        combine_denoised_hijack,
-        original_function=sampler.model_wrap_cfg.combine_denoised
+        denoised_uncond = denoised[-1]
+        denoised_conds = torch.stack(denoised[:-1], dim=0)
+        denoised_cond = torch.mean(denoised_conds, dim=0)
+        forge_denoised = denoised_uncond + (denoised_cond - denoised_uncond) * cond_scale
+
+        cond_indices = cond_composition[0]
+        prompt = global_state.prompt_exprs[0]
+
+        for batch_i in range(denoised_uncond.shape[0]):
+            args = CombineDenoiseArgs(denoised_conds.unbind(dim=1)[batch_i], denoised_uncond[batch_i], cond_indices)
+            cond_delta = prompt.accept(CondDeltaVisitor(), args, 0)
+            aux_cond_delta = prompt.accept(AuxCondDeltaVisitor(), args, cond_delta, 0)
+            cfg_cond = forge_denoised[batch_i] + aux_cond_delta * cond_scale
+            forge_denoised[batch_i] = cfg_rescale(cfg_cond, denoised_uncond[batch_i] + cond_delta + aux_cond_delta)
```

Comment:
the cfg_rescale could be dropped here I think, as forge has built-in extensions with more control over it

Reply:
Very good point, we should also hide the rescale slider in the ui if forge is detected.